src/bpf/bpfd-manager.c

```diff
@@ -0,0 +1,489 @@
+        /* Iterate through all BPF maps to get map names, map types, and memlock */
+        while (sym_bpf_map_get_next_id(id, &id) >= 0) {
+                _cleanup_close_ int fd = -EBADF;
+                _cleanup_free_ char *memlock = NULL;
+                char *map_name = NULL;
+                uint32_t map_id = id;
+                JsonVariant *v;
+
+                zero(map_info);
+                /* Grab the FD using the ID*/
+                fd = sym_bpf_map_get_fd_by_id(id);
+                if (fd < 0) {
+                        log_error("Can't get fd for map id: %u", id);
+                        continue;
+                }
+
+                /* Fetch bpf_map_info using fd */
+                r = sym_bpf_obj_get_info_by_fd(fd, &map_info, &map_info_len);
+                if (r < 0) {
+                        log_error_errno(r, "Can't get bpf map info for fd %i: %m", fd);
+                        continue;
+                }
+
+                /* Store map names into hashmap for later lookup when iterating progs */
+                map_name = strdup(map_info.name);
+                if (!map_name)
+                        return -ENOMEM;
+                hashmap_put(map_name_by_id, UINT32_TO_PTR(map_id), map_name);
+
+                /* Fetch memlock from /proc/self/fdinfo using map fd */
+                r = get_memlock_by_fd(fd, &memlock);
+                if (r < 0) {
+                        log_error_errno(r, "Failed to read map memlock from /proc/self/fdinfo/%i: %m", fd);
+                        continue;
+                }
+
+                if (link) {
+                        r = json_build(&v, JSON_BUILD_OBJECT(
+                                                JSON_BUILD_PAIR_STRING("type", "map"),
+                                                JSON_BUILD_PAIR_OBJECT("data",
+                                                JSON_BUILD_PAIR_STRING("name", map_info.name),
+                                                JSON_BUILD_PAIR_UNSIGNED("type", map_info.type),
+                                                JSON_BUILD_PAIR_STRING("memlock", memlock))));
+                        if (r < 0) {
+                                log_error_errno(r, "Failed to build JSON object: %m");
+                                continue;
+                        }
+
+                        r = varlink_notify(link, v);
+                        if (r < 0) {
+                                log_error_errno(r, "Failed to notify varlink: %m");
+                                continue;
+                        }
+                } else {
+                        log_struct(LOG_INFO,
+                                LOG_MESSAGE("Logging info about BPF map: %s", map_info.name),
+                                "MESSAGE_ID=" SD_MESSAGE_BPFD_LOG_STR,
+                                "NAME=%s", map_info.name,
+                                "TYPE=%u", map_info.type,
+                                "MEMLOCK=%s", memlock);
```

Comment:
Instead of logging a single line per BPF program/map we should log the JSON object containing the data on all maps and programs once. We'd construct the same JSON object whether we're doing varlink or logging to the journal, the only difference should be where we write the JSON object to.